Add/805-playlist-refactor-2

antpb opened the pull request:
This is a clean PR to add the playlist block to Gutenberg. There were merge conflicts that resulted in me needing to create a new branch based on the newest version of Gutenberg. Previous conversation can be found here: [#7126](https://github.com/WordPress/gutenberg/pull/7126/)

There is one bug currently with notices where the mismatched type error shows 4 of the same notice. Working through that and we should be good for another review. 

CC: @karmatosed @noisysocks

antpb committed:
initial commit

initial commit

antpb committed:
playlist init

starting playlist over

antpb committed:
storing ids in comment as ids attribute. moving to render callback....may need core mod >_<

storing ids in comment as ids attribute. moving to render callback....may need core mod >_<

antpb committed:
storing ids in comment as ids attribute. moving to render callback....may need core mod >_<

storing ids in comment as ids attribute. moving to render callback....may need core mod >_<

antpb committed:
playlist_block_callback_dependencies function. adds index.php

Commit:
cleaning up attributes and logging code. removes mediaelement component

Commit:
serversiderender component being used now in edit...workign through attributes error

Commit:
removes client-asset changes meant for mediaelement component

Commit:
adds media-upload changes to support playlist modal

Commit:
adds type attribute for video/audio conditional for the render callback

antpb committed:
fixed edit state and now when you edit post, it acts as expected - commits from airport terminals ✈

antpb committed:
Adjusted some of the wording to include video language...not just audio

antpb committed:
Group coding by antpb - noisysocks - soean

antpb committed:
Fixing noisysocks suggested changes

joemcgill committed:
Load playlist block alphabetically

This simply moves the playlist block in the list of blocks imported and registered in core-blocks/index.js.

joemcgill committed:
Clean up lint errors

This commit generally accomplishes a few thing:
- Removes any unused variables and functions
- Ensuring all used functions and variables are declared
- Simplifies some variable declarations
- Fixing spacing to conform to WP style

joemcgill committed:
Clean up dynamic rendering PHP code

This fixes up code style issues, adds correct docblocks, and removes unneccessary script rendering which isn't necessary during server rendering.

joemcgill committed:
Wrap server rendered playlists in figure element

This wraps the shortcode output in a `figure` element with a class of `wp-block-playlist` applied so we can style the block accordingly.

joemcgill committed:
Remove non-playlist styles from the playlist stylesheet.

joemcgill committed:
Fix playlist margins in the editor

This adds an editor.scss file to the playlist block that removes margins from `.wp-playlist-block` elements, to avoid margins being added as a default style by browsers.

joemcgill committed:
Fix alignment support for the playlist block

This adds an align attribute to the playlist block and adds the proper alignment class to the markup on the front end to support alignments. This also deletes the style.scss file, which now seems unnecessary.

antpb committed:
Merge pull request #1 from joemcgill/add/playlist-block

Playlist block updates @joemcgill  fixed a ton of formatting and bugs. :D

antpb committed:
Fixed edit button to auto populate media modal

antpb committed:
cleaned up spacing issues

antpb committed:
fixes Soean's suggestions and adding attributes to fix settings issue

antpb committed:
removes some silly test code

Soean committed:
Add controls and fix lint errors

antpb committed:
Merge branch 'master' into add/playlist-block-805

antpb committed:
removes `SettingsView` from Playlist Media view

antpb committed:
fixes gallery mutiple bug - Nice find @noisysocks !

antpb committed:
sets defaults to on for tracklist, images, artist as per Core defaults

antpb committed:
sets defaults for front end render as well as fixes lint errors (someone finally installed a linter!)

antpb committed:
removes silly array of arrays and adds back in SettingsView to make backbone happy

antpb committed:
adds style control for playlist: dark, light

antpb committed:
should fix merge conflict

antpb committed:
changes edit-post/hooks/blocks to components for merge conflict

antpb committed:
fixes more merge conflicts

antpb committed:
Merge branch 'master' into add/playlist-block-805

antpb committed:
uses new IconButton component to match gallery

antpb committed:
Merge branch 'add/playlist-block-805' of https://github.com/antpb/gutenberg into add/playlist-block-805

antpb committed:
fixes `ServerSideRender` update for settings

antpb committed:
@soean provided me the golden snippet of code to fix uploads

antpb committed:
removes `SettingsView`

antpb committed:
linting fix from spaces to tabs

Soean committed:
Fixes WPCS warnings

Soean committed:
Remove isPlainObject and map from SSR

antpb committed:
fixes `Media Library` button

antpb committed:
auto sets type based on first item in media array passed back from uploads and  media selection

antpb committed:
adds type attribute to edit button to pass back to the media views...need to add condition in media view for type

antpb committed:
adds edit condition for videos so media view focuses on video's and not a default of audio

antpb committed:
changes frame to select since that is what a user would probably want to do

antpb committed:
adds @karmatosed design recommendations and fixes an upload bug introduced

when uploading, we are assuming that the type is audio by default because for whatever reason the upload form is not passing back the mime type

Soean committed:
Improve toggle functions

antpb committed:
hides settings in media modal on gutenberg editor pages. ¯\_(ツ)_/¯

Soean committed:
Fix toggle style

antpb committed:
fixing merge conflict

antpb committed:
fixing merge issue of including package-lock

antpb committed:
updating package-lock to match master because my local environment was silly

antpb committed:
Merge branch 'master' into add/playlist-block-805

antpb committed:
final changes to remove package-lock.json from this PR

antpb committed:
removing extra =

Soean committed:
Alow upload video and audio files

Soean committed:
Refactor Code

Soean committed:
Add tests

antpb committed:
changes playlist icons to better indicate both video AND audio.

antpb committed:
Revert "changes playlist icons to better indicate both video AND audio."

This reverts commit 4495b53d1375a4be9df3dffb0339526f0a4e58b3.

antpb committed:
refactor naming of editing state and code tweak

antpb committed:
fixes conditional in onFileChange

antpb committed:
refactors naming of addFiles to onUploadFiles and removes unnecessary attributes

antpb committed:
adds period to Placeholder instructions

antpb committed:
fixes accepted types to comma-separated string

antpb committed:
fixes style options to be inline. maybe more styles some day :D

antpb committed:
sets type default to audio to initialize media modal

antpb committed:
sets mediaUpload value correctly using `mediaIds`

antpb committed:
spaces child of component correctly

antpb committed:
removes lint rules

antpb committed:
css spacing

antpb committed:
simplifies margin css

antpb committed:
renames artists attribute to better indicate action

antpb committed:
fixes typo

antpb committed:
fixes sneaky tab

antpb committed:
fixes other sneaky tab

antpb committed:
renames `playlistItemObject` to `slimPlaylistItemObject`

antpb committed:
fixes abbreviations

antpb committed:
fixes strcuture of `multiple` and `playlist` checks in mediaupload

antpb committed:
fixes slimPlaylistItemObject

antpb committed:
adds castArray to utils/mediaupload

antpb committed:
Merge branch 'master' into add/playlist-block-805

antpb committed:
fixes `showArtists` in index.php

antpb committed:
sets default for php render to default 'audio' fixes a bug introduced with defaults. :)

antpb committed:
adds space again to css file

antpb committed:
fixes upload issue adding back in mime_type removed from merge conflict

antpb committed:
renames mime_type to mimeType

antpb committed:
removes test code for another issue...

antpb committed:
sets attribute 'artists' as 'showArtists' as the shortcodes expects. Also sets work for notice on mismatch type

antpb committed:
adds showTrackNumbers attribute to expose more of the shortcode options

antpb committed:
uses this.props.attributes to fix edit state bug.

For some reason passing attributes to the constructor does not allow the edit state to be determined. reverting to this.props.attributes fixes the issue

antpb committed:
changes playlist block description to exclude `core` verbiage

antpb committed:
Merge branch 'add/playlist-block-805' into add/805-playlist-refactor

antpb committed:
php coding standards fix

ianbelanger79 committed:
Added more tabs to comply with coding standards

Added 2 tabs to array args to bring them in "front" of their arrays

ianbelanger79 committed:
Moved => inline with the rest

antpb committed:
First pass at error notices on inconsistent type

antpb committed:
should use proper mediaUpload function. attempting to refresh master...lots of merge conflicts

antpb committed:
Merge branch 'add/805-playlist-refactor' into add/8050-playlist-refactor-2

antpb committed:
fixes css and adopts new block structure

antpb committed:
updates code to new standards in latest Gutenberg while fixing Notices and Uploads

antpb committed:
removes some from lodash because we dont want THAT some :)

antpb committed:
fixes fixture issue Travis CI identified

antpb committed:
fixes snapshots for Travis

antpb committed:
fixes more fixture things for tests

antpb committed:
trying to make phpcs happy

antpb committed:
fixes linting issue with blockEditRender now properly being imported

antpb committed:
attempting to appease Travis

antpb committed:
fixing blockEditRender path and bringing php function to one line

antpb committed:
fixes eslint issue

antpb committed:
adding hasError state to help with duplicate notices being thrown for mismatched type

noisysocks commented:
There are some linting errors which is causing the Travis CI build to fail. You can run `npm run lint` locally to see a list of the errors.

> There is one bug currently with notices where the mismatched type error shows 4 of the same notice. Working through that and we should be good for another review.

👌 ping me when you're ready!

antpb committed:
takes condition out of render and adds componentDidUpdate to trigger the mismatched type notice

anthonyburchell commented:
Hi @noisysocks ! I have updated things to solve some of the Travis CI stuff. We have on failure still on the php side of thins. I noticed some weirdness in the php checks. It says that I have linting errors but running it locally `npm run lint` does not produce any issues. 

Lint results from Travis below:

```
PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
----------------------------------------------------------------------
    SOURCE                                                       COUNT
----------------------------------------------------------------------
[x] PEAR.Functions.FunctionCallSignature.MultipleArguments       1
----------------------------------------------------------------------
A TOTAL OF 1 SNIFF VIOLATION WERE FOUND IN 1 SOURCE
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SOURCES AUTOMATICALLY (1 VIOLATIONS IN TOTAL)
----------------------------------------------------------------------
```

Aside from that, notices seem to be working just fine, but we are still getting duplicates. I've broken out the trigger for notice to a `componentDidUpdate` method checking the `hasError` state, however, I am still seeing the notice duplicate three times on mismatched type (one less duplicate than before!) Would love to have some eyes on it to see what I'm missing. :)

antpb committed:
moves resetting the state to onSelectMedia. Now producing one notice per file uploaded

antpb committed:
removes error state from else condition on consistent type check

anthonyburchell commented:
Update: I've changed the way that states are reset and triggered. This has brought down the duplicate errors to something that makes a little more sense. It still seems to duplicate the notice for each error file uploaded...I'm working through that now to make sure we only do it once.

antpb committed:
fixes bug where any upload produced error message regardless of type check

antpb committed:
moves first type and isConsisstentType const outside of the onFileChange method to ensure it is only set once

antpb committed:
adds suggestions on setting state

antpb committed:
fixes mismatched type issues by changing condition to check type and not mimeType

antpb committed:
fixes type check

aduth commented on packages/block-library/src/playlist/index.js:
```diff
@@ -0,0 +1,65 @@
+
+	attributes: {
+		ids: {
+			type: 'string',
```
Should this be a string? Or an `array`?

anthonyburchell replied:
with your help this is no longer an issue! :D thanks again!

antpb committed:
fixes state issue where null was accepted as ids

antpb committed:
fixes weirdness where we needed to stringify and json encode the ids string!  woot!

antpb committed:
fixes tests for new ids array

antpb committed:
fixes tests for ids array

antpb committed:
fixes js tests

antpb committed:
removes mediaIds as it is no longer needed

antpb commented:
Update here: Remaining bugs have been fixed and the code has been cleaned up. We've removed the `hasError` state and are throwing the notice of mismatched types only once. With @aduth's guidance I was able to get `ids` to be accepted as an array. Previously this caused issues on the front end render of blocks.  Attributes in `index.php` needed the `items` type defined as numbers. This fixes the weirdness with REST API stuff found early in the building of this block.

I've updated tests to use the new array. These seem to pass Travis CI. I have one remaining Travis CI check that is failing. It is claiming MultipleArguments failure in index.php. I am unable to replicate this failure locally.

```[x] PEAR.Functions.FunctionCallSignature.MultipleArguments       1```

youknowriad commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,204 @@
+
+		if ( isEditing ) {
+			return (
+				<Placeholder
```
Is it possible to use the `MediaPlaceholder` instead of a custom one like proposed in #9139 This will ensure the media blocks all have the same "design". Or is this very specific?

anthonyburchell replied:
I think this one is specific due to the type validation in the upload form. I could be wrong. I'll see if I can match the MediaPlaceholder arguments. Also, I may need to add some new conditions for `playlist` into MediaPlaceholder... for the sake of getting this merged before we hit conflicts could this be a future issue to fix in a later iteration? I've hit a few merge blockers that cause me a bit of time getting updated. Would love to have a baseline to build off of.

youknowriad replied:
I'm fine for this being a follow-up PR :), but I do think we should try to do it for the sake of consistency.

aduth commented on packages/editor/src/utils/media-upload/media-upload.js:
```diff
@@ -68,7 +68,7 @@ export function mediaUpload( {
 
 	// Allowed type specified by consumer
 	const isAllowedType = ( fileType ) => {
-		return ( allowedType === '*' ) || startsWith( fileType, `${ allowedType }/` );
```
Aren't we breaking support for the `'*'` wildcard by removing it here? Was that in use anywhere, or documented? Also saddening if this is not breaking any tests with its removal, since it means it is not well-covered by tests.

anthonyburchell replied:
If `*` was the wild card desired couldn't that be defined in the allow type argument of mediaPlaceholder as `allowedType` which would get passed here? Reverting this line causes successful type check uploads to fail for Playlists...ill have to dig into that if we want to revert this. the Files block uses `*` and I have been testing that on in tandem with making the playlist block. Thoughts?

aduth replied:
I'm not really sure I understand.

- What failures were you seeing without these changes? You're not specifying `*` so I don't see why it would have any impact to leave it.
- Regardless of whether a `*` is sensible, it exists and is already in use by the Files block. If we want to remove it, we need to remove any references to it. Speaking personally, I don't really like `*` as the wildcard. It seems like if someone wants any file to be uploaded, they should just omit the `allowedType` property altogether. But then again, this change has nothing to do with a Playlist block implementation and seems like it should be a separate effort.

antpb replied:
I've reverted this line and uploads no longer work for playlist only...going to have to dig in to figure out whats going wrong. I cant even get a logging of `fileType` now. For what it's worth, the file block was accepting all types previously before reverting the change

aduth commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	initializePlaylist() {
+		window.wp.playlist.initialize();
```
Noting that globals should be considered "evil" in this context. We want to eliminate any direct references to `window.wp` with an equivalent import if it's possible. I'm sensing that for playlists specifically, this may be difficult to accomplish. Since `block-library` is a [published module](https://www.npmjs.com/package/@wordpress/block-library), we should probably be conscious of any global dependencies we assume to exist by documenting them. Even in the context of WordPress, it's not obvious on the surface that we need playlist scripts to be available any time we load the `block-library` script.

aduth commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	initializePlaylist() {
+		window.wp.playlist.initialize();
```
Given the behavior of `ServerSideRender` to trigger this on any update to the response, do we need to consider the fact that the rendered preview could be updated multiple times? Are we leaving [zombie event handlers](https://lostechies.com/derickbailey/2011/09/15/zombies-run-managing-page-transitions-in-backbone-apps/) by not cleaning up after the old content? I'm not entirely sure how `wp.playlist.initialize` is intended to work here.

anthonyburchell replied:
This is needed because the initialization needs to happen after the editor enqueues the dependencies listed in client-assets here: https://github.com/WordPress/gutenberg/pull/9169/files#diff-620130744b6b73b822a3df609671d930

A playlist will not render correctly (using mejs) unless the initialize happens at this point. I worked a bit with @noisysocks on this until we determined this is where the initialization should be called to fix the render

aduth commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	uploadFromFiles( event ) {
+		this.onUploadFiles( event.target.files );
```
Curious: Is there ever a chance this can be called with an empty `event.target.files` ? If so, how would we handle that?

anthonyburchell replied:
I can wrap this in a condition if it exists. I cant think of an instance where this would happen as upload from files is triggered after files are selected. Worth the condition I think

aduth replied:
Being defensive is good, but having an understanding of the types of values we're expecting to handle is better.

anthonyburchell commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,204 @@
+		this.uploadFromFiles = this.uploadFromFiles.bind( this );
+		this.onUploadFiles = this.onUploadFiles.bind( this );
+
+		// check for if ids is set to determine edit state
```
I was using the Audio block as reference when making this state which also uses a `isEditing`. If we do it here we should probably do it on that block for consistency.

aduth replied:
I agree.

anthonyburchell commented on packages/block-library/src/playlist/editor.scss:
```diff
@@ -0,0 +1,13 @@
+	border: none;
+}
+
+.gutenberg-editor-page .playlist-settings {
```
This is so we don't add padding to regular playlist shortcodes that are not using Gutenberg. These are just some padding changes that were asked from Tammie in an earlier Design review.

aduth replied:
But `editor.css` is intended to be used only in the context of the block editor anyways?

anthonyburchell replied:
I'll verify with a classic editor / Gutenberg hybrid instance and see if the class impacts the classic content.

anthonyburchell commented on packages/block-library/src/playlist/index.js:
```diff
@@ -0,0 +1,65 @@
+	edit,
+
+	save() {
+		return null;
```
I agree. In the initial creation of this block I went the route of static but we needed things like track length, looping, auto play of next track to be available. `mediaelement.js` is what allows for those things. I went so far as to try and build a `mediaelement` component to handle that stuff. It went badly and I spent months on it before I was told we should go the server side render route. Im very conscious of the fact that we need a future iteration that is more dynamic. The way this playlist block differs from the current experience is that it exposes all options in the sidebar aside from just a few options in the creation of a playlist in the media modal. A start

anthonyburchell commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,204 @@
+						multiple
+						className="wp-block-playlist__upload-button"
+						onChange={ this.uploadFromFiles }
+						accept="audio/*,video/*"
```
The uploads do not succeed if a mime type that is not allowed is uploaded. Just tested this. It does not produce a visible error though. This is also true for the file, audio, and video blocks. It is caught in the `allowedMimeTypes` and `allowedMimeTypesForUser`. We'll probably want a new issue for this not producing a notice. I'm also not sure how we would trigger the notice in `mediaUpload` 

(I tested using a .aac audio file)

anthonyburchell commented on packages/block-library/src/playlist/index.js:
```diff
@@ -0,0 +1,65 @@
+
+	category: 'common',
+
+	attributes: {
```
I might need some clarification here. Are you saying this is a duplicate of the attributes in the register block function in `index.php`? To test that theory, I removed the attributes and now my `ids` condition is not passing for the edit state (no longer a state but a condition check ;) )

aduth replied:
I tested locally and I see no obvious issues when removing `attributes` from this JavaScript file. Can you elaborate on what you're seeing?

antpb replied:
Sorry. Disregard. I may have had something else broken when I initially tested this. I've pushed the change to remove attributes from index.js

anthonyburchell commented on packages/block-library/src/playlist/index.php:
```diff
@@ -0,0 +1,71 @@
+	$attributes['artists']      = $attributes['showArtists'];
+	$attributes['tracknumbers'] = $attributes['showTrackNumbers'];
+
+	$classes = empty( $attributes['align'] ) ? 'wp-block-playlist' : 'wp-block-playlist align' . $attributes['align'];
```
Cool. Agreed. thanks

anthonyburchell commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,204 @@
+		const { isEditing } = this.state;
+		const { tracklist, showTrackNumbers, showArtists, images, style, type } = attributes;
+
+		const onSelectMedia = ( media ) => {
```
cool. I just got that moved out properly.

antpb committed:
refactor based on review by @aduth

antpb committed:
removes isEditing state completely

anthonyburchell commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,204 @@
+		};
+	}
+
+	initializePlaylist() {
```
This was intended to self document what it was doing. I'll add some documentation to the function. I added some docs that explain that this is used to initialize after the editor has enqueued dependencies

antpb committed:
adds initializePlaylist documentation

antpb committed:
pulls getMimeBaseType out of the component

anthonyburchell commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,204 @@
+
+	onUploadFiles( files ) {
+		function getMimeBaseType( type ) {
+			return type.split( '/' )[ 0 ];
```
I've added a condition to check if type is passed. If it is not defined it will throw a notice of 'MIME Type could not be determined'

anthonyburchell replied:
Not sure the best way to test that is working...if you have any suggestions it would be appreciated :)

antpb committed:
adds condition to verify type is passed from the files uploaded

antpb committed:
adds check to if files list exists

antpb committed:
fixes JS tests

aduth commented on packages/components/src/server-side-render/index.js:
```diff
@@ -44,10 +44,15 @@ export class ServerSideRender extends Component {
 		if ( ! isEqual( prevProps, this.props ) ) {
 			this.fetch( this.props );
 		}
+		if ( this.state.response !== prevState.response ) {
```
It's a bit awkward to call a prop in response to a state change in `componentDidUpdate`, since a component manages its own state and begs the question why isn't the prop called at the time we're calling `setState` ? Our current implementation of `ServerSideRender#fetch` doesn't make this very easy to refactor, though...

antpb replied:
Is this a merge blocker? If not can we bring it to an issue after? I think there would have to be an alteration of ServerSideRender to accomplish this. I fear it may introduce more bugs

antpb committed:
removes attributes from index.js to make things tidy

antpb committed:
fixes comment spacing for JS standards

antpb commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,214 @@
+	onSelectMedia( media ) {
+		const { setAttributes } = this.props;
+		//check if there are returned media items and set attributes when there are
+		if ( media && media[ 0 ].url ) {
```
I don't believe it will ever be unless there is an error maybe. We also have a similar check in the Audio block. I was matching that for consistency on the chance there was an instance I was unaware of: https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/audio/edit.js#L88

antpb committed:
removes condition

antpb commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,214 @@
+	}
+
+	/**
+	* The function is used after the editor has enqueued playlist dependencies.
```
fixed :)

antpb commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,214 @@
+	onSelectMedia( media ) {
+		const { setAttributes } = this.props;
+		//check if there are returned media items and set attributes when there are
+		if ( media && media[ 0 ].url ) {
```
we shouldnt. I've removed the condition

antpb committed:
fixes error message capitalization

antpb commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,214 @@
+	getMimeBaseType( type ) {
+		const { noticeOperations } = this.props;
+		if ( ! type ) {
+			noticeOperations.createErrorNotice( 'MIME Type could not be determined' );
```
I've fixed the capitalization. Will think through the multiple messages bit. Can create an issue for it also after merge if we want to save that for future work.

antpb committed:
adds localization, fixes gallery view defaults, and fixes upload issues for '*'

gziolo commented on packages/block-library/src/playlist/edit.js:
```diff
@@ -0,0 +1,212 @@
+	*/
+
+	initializePlaylist() {
+		window.wp.playlist.initialize();
```
@youknowriad, this block depends on both WordPress global `wp.playlist` and `ServerSideRender` component which is very backend tied. We need to do something about it. There should be reusable blocks which work everywhere and blocks which are reusable but WordPress specific. If we would draw this distinction we wouldn't have to worry about `Code` and `Freeform` anymore.

youknowriad replied:
yes, a first step would be to move it ouside of the packages like `html` or `freeform`. Later on, we'd need to rewrite `wp.playlist` as a Gutenberg package and then move the block to the packages as well.

gziolo commented on lib/client-assets.php:
```diff
@@ -1357,6 +1357,8 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 		'post' => $post->ID,
 	) );
 	wp_enqueue_editor();
+	wp_playlist_scripts( 'audio' );
```
Shouldn't `wp.playlist` to be listed as a dependency of `wp-block-library`? It's referenced it in the code so I would assume that it needs to be.

gziolo commented:
Did you check how much more data users will have to load in their browser to be able to use this block? In particular, how much weigh all the files enqueued with those two lines:
```js
wp_playlist_scripts( 'audio' );
wp_playlist_scripts( 'video' );
```

antpb commented:
@gziolo there was some previous discussion about this in an earlier review. It adds about 110-120kb : https://github.com/WordPress/gutenberg/pull/7126#discussion_r199875612

aduth commented:
The additional weight is significant, particularly for a block which won't always be commonly used by a large subset of users.

I wonder if it's something where we should employ the technique as we did for the HTML block and CodeMirror:

https://github.com/WordPress/gutenberg/blob/eac95b5638db8f7d9aba8f3ddf0cd55138edbb81/components/code-editor/index.js#L28-L58

Also related: #8017

aduth commented:
Raising other options to explore / consider:

- Simplify what we're displaying as a preview in the editor. Avoid pulling in any dependencies and just render a simplified list of file names (no in-browser preview in editor).
- Make it the responsibility of ServerSideRender to bundle the relevant scripts, whether that's inline or as part of an embedded iframe.

antpb commented:
@aduth After digging into this a bit today, I remember the thought process in adding the playlist enqueues to client-assets. If you remove the script calls from that file and create a playlist block, the block responds that it has encountered an error and cannot load. Saving/publishing the post and refreshing the editor magically allows it to render with it's dependencies. 

Something is not triggering in ServerSideRender at the right time to enqueue the dependencies. Still digging through this, but I think the best path forward is the lazy load solution proposed earlier with a state that says that it has dependencies before render.

joemcgill commented:
Related to #805

antpb committed:
conditionally loads playlist dependencies instead of doing it in client-assets.php

gziolo commented:
@melchoyce @mapk @jasmussen @karmatosed @youknowriad - what's the plan regarding this block? It isn't listed as the block planned for phase 2.

mapk commented:
> what's the plan regarding this block? It isn't listed as the block planned for phase 2.

Definitely not on our list of priorities right now. 

I'm going to close this because it hasn't been touched since September 2018. But I don't want to discourage contributors from working on something they're passionate about. If anyone feels like refactoring this and opening it back up, go ahead. 😄
